sui/token_bridge/sources/state.move

```diff
@@ -150,34 +150,42 @@ module token_bridge::state {
         self: &mut State,
         receipt: UpgradeReceipt
     ): ID {
-        // Uptick the upgrade cap version number using this receipt.
-        package::commit_upgrade(&mut self.upgrade_cap, receipt);
-
-        // Check that the upticked hard-coded version version agrees with the
-        // upticked version number.
+        // Check that the hard-coded version version agrees with the version
+        // number in the `UpgradeCap`. We should only be allowed to upgrade
+        // using the latest build.
         assert!(
-            package::version(&self.upgrade_cap) == control::version() + 1,
+            package::version(&self.upgrade_cap) == control::version(),
             E_BUILD_VERSION_MISMATCH
         );
 
+        // Uptick the upgrade cap version number using this receipt.
+        package::commit_upgrade(&mut self.upgrade_cap, receipt);
+
         // Update global version.
         required_version::update_latest(
             &mut self.required_version,
             &self.upgrade_cap
         );
 
-        // Enable `migrate` to be called after commiting the upgrade.
+        // Require that `migrate` be called only from the current build.
+        require_current_version<control::Migrate>(self);
+
+        // We require that a `MigrateTicket` struct be destroyed as the final
+        // step to an upgrade by calling `migrate` from the `migrate` module.
         //
         // A separate method is required because `state` is a dependency of
         // `migrate`. This method warehouses state modifications required
         // for the new implementation plus enabling any methods required to be
         // gated by the current implementation version. In most cases `migrate`
-        // is a no-op. But it still must be called in order to reset the
-        // migration control to `false`.
+        // is a no-op.
+        //
+        // The only case where this would fail is if `migrate` were not called
```

Comment:
I would be careful here. `migrate` must be called in a separate transaction, which means the `upgrade` + `migrate` flow is not atomic, which opens up the possibility that `upgrade` succeeds, but `migrate` fails for some reason. An example of a failure that would not come up in local testing is if there's a loop that iterates over some state that with mainnet data exceeds the maximum transaction budget. If the `migrate` call fails, then further upgrades become impossible. We need to be able to recover from such scenarios

Reply:
It wasn't clear that I changed `migrate` to be optional, so I amended the commit where I added the `MigrateTicket`: https://github.com/wormhole-foundation/wormhole/pull/2753/commits/0997424c11afef6a5e53b74f4118e5cb865e37c7.